File: python/template/services/service_test.py
import asyncio
import pytest
from unittest.mock import AsyncMock
from models import (
{%- for resource in resources %}
    {{ resource.name }},
    {{ resource.name }}Response,
{%- endfor %}
)
from services import (
{%- for resource in resources %}
    {{ resource.name }}Service,
{%- endfor %}
)
from repositories import (
{%- for resource in resources %}
    {{ resource.name }}Repository,
{%- endfor %}
)

{% for resource in resources %}
{%- set r = resource.name %}
{%- set slug = r | to_snake %}
_{{ slug }}_responses = [
    {{ r }}Response(id='ac1df01c-7ece-4a20-ab60-179829dad8f5', name='mockName1', type='mockType1'),
    {{ r }}Response(id='de6cbc87-5969-458c-8444-3512a82250bc', name='mockName2', type='mockType2')
]
_{{ slug }}_update_response = {{ r }}Response(id='ac1df01c-7ece-4a20-ab60-179829dad8f5', name='mockName1-Update', type='mockType1-Update')


def describe_{{ slug }}_service():
    @pytest.fixture
    def mock_repository():
        mock = AsyncMock({{ r }}Repository)
        mock.get_by_id.return_value = _{{ slug }}_responses[0]
        mock.get_list.return_value = _{{ slug }}_responses
        mock.create.return_value = _{{ slug }}_responses[0]
        mock.update.return_value = _{{ slug }}_update_response
        mock.replace.return_value = _{{ slug }}_update_response
        return mock

    @pytest.fixture
    def service(mock_repository):
        return {{ r }}Service(repository=mock_repository)
{%- if "get_by_id" in resource.operations %}

    def describe_get_by_id():
        def test_successfully_call_repository(service, mock_repository):
            result = asyncio.run(service.get_by_id(item_id='ac1df01c-7ece-4a20-ab60-179829dad8f5'))
            mock_repository.get_by_id.assert_called_once()
            assert result == _{{ slug }}_responses[0]
{%- endif %}
{%- if "list" in resource.operations %}

    def describe_get_list():
        def test_successfully_call_repository(service, mock_repository):
            result = asyncio.run(service.get_list())
            mock_repository.get_list.assert_called_once()
            assert result == _{{ slug }}_responses

        def test_passes_limit_through(service, mock_repository):
            asyncio.run(service.get_list(limit=25))
            mock_repository.get_list.assert_called_once_with(25)
{%- endif %}
{%- if "create" in resource.operations %}

    def describe_create():
        def test_successfully_call_repository(service, mock_repository):
            mock_item = {{ r }}(name='mockName1', type='mockType1')
            mock_item.id = 'ac1df01c-7ece-4a20-ab60-179829dad8f5'
            result = asyncio.run(service.create(item=mock_item))
            mock_repository.create.assert_called_once()
            assert result == _{{ slug }}_responses[0]
{%- endif %}
{%- if "update" in resource.operations %}

    def describe_update():
        def test_successfully_call_repository(service, mock_repository):
            mock_update_item = {{ r }}(id='ac1df01c-7ece-4a20-ab60-179829dad8f5', name='mockName1-Update', type='mockType1-Update')
            result = asyncio.run(service.update(item=mock_update_item))
            mock_repository.update.assert_called_once()
            assert result == _{{ slug }}_update_response
{%- endif %}
{%- if "replace" in resource.operations %}

    def describe_replace():
        def test_successfully_call_repository(service, mock_repository):
            mock_replace_item = {{ r }}(id='ac1df01c-7ece-4a20-ab60-179829dad8f5', name='mockName1-Update', type='mockType1-Update')
            result = asyncio.run(service.replace(item=mock_replace_item))
            mock_repository.replace.assert_called_once()
            assert result == _{{ slug }}_update_response
{%- endif %}
{%- if "delete" in resource.operations %}

    def describe_soft_delete():
        def test_successfully_call_repository(service, mock_repository):
            asyncio.run(service.soft_delete(item_id='ac1df01c-7ece-4a20-ab60-179829dad8f5'))
            mock_repository.delete.assert_called_once()
{%- endif %}

{% endfor %}

File: go/template/services/service_test.go
package services

import (
	"context"
	"testing"

	"{{project_endpoint}}/models"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)
{% for resource in resources %}
{%- set r = resource.name %}
type mock{{ r }}Repository struct {
	mock.Mock
}

func (m *mock{{ r }}Repository) Get(ctx context.Context, id string) (*models.{{ r }}Dto, error) {
	args := m.Called(ctx, id)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.{{ r }}Dto), args.Error(1)
}

func (m *mock{{ r }}Repository) GetList(ctx context.Context, limit int) ([]models.{{ r }}Dto, error) {
	args := m.Called(ctx, limit)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).([]models.{{ r }}Dto), args.Error(1)
}

func (m *mock{{ r }}Repository) Create(ctx context.Context, item models.{{ r }}) (*models.{{ r }}Dto, error) {
	args := m.Called(ctx, item)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.{{ r }}Dto), args.Error(1)
}

func (m *mock{{ r }}Repository) Update(ctx context.Context, item models.{{ r }}) (*models.{{ r }}Dto, error) {
	args := m.Called(ctx, item)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.{{ r }}Dto), args.Error(1)
}

func (m *mock{{ r }}Repository) Replace(ctx context.Context, item models.{{ r }}) (*models.{{ r }}Dto, error) {
	args := m.Called(ctx, item)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.{{ r }}Dto), args.Error(1)
}

func (m *mock{{ r }}Repository) Delete(ctx context.Context, id string) error {
	args := m.Called(ctx, id)
	return args.Error(0)
}

func setup{{ r }}ServiceTest() (*mock{{ r }}Repository, {{ r }}Service) {
	mockRepo := new(mock{{ r }}Repository)
	service := New{{ r }}Service(mockRepo)
	return mockRepo, service
}
{%- if "get_by_id" in resource.operations %}

func TestGet{{ r }}_ShouldReturn{{ r }}Dto(t *testing.T) {
	mockRepo, service := setup{{ r }}ServiceTest()
	id := "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c"
	expected := &models.{{ r }}Dto{Id: id, Name: "mock{{ r }}"}
	mockRepo.On("Get", mock.Anything, id).Return(expected, nil)

	result, err := service.Get(context.Background(), id)

	assert.NoError(t, err)
	assert.Equal(t, expected, result)
	mockRepo.AssertExpectations(t)
}
{%- endif %}
{%- if "list" in resource.operations %}

func TestGet{{ r }}List_ShouldReturnListOf{{ r }}Dto(t *testing.T) {
	mockRepo, service := setup{{ r }}ServiceTest()
	expected := []models.{{ r }}Dto{
		{Id: "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c", Name: "mock{{ r }}1"},
		{Id: "5615ff05-3032-4459-88ad-b6a4c3e51ca0", Name: "mock{{ r }}2"},
	}
	mockRepo.On("GetList", mock.Anything, 50).Return(expected, nil)

	result, err := service.GetList(context.Background(), 50)

	assert.NoError(t, err)
	assert.Equal(t, expected, result)
	mockRepo.AssertExpectations(t)
}
{%- endif %}
{%- if "create" in resource.operations %}

func TestCreate{{ r }}_ShouldReturnCreated{{ r }}Dto(t *testing.T) {
	mockRepo, service := setup{{ r }}ServiceTest()
	createRequest := models.Create{{ r }}Request{Name: "mockCreate{{ r }}"}
	expected := &models.{{ r }}Dto{Id: "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c", Name: "mockCreate{{ r }}"}
	mockRepo.On("Create", mock.Anything, mock.AnythingOfType("models.{{ r }}")).Return(expected, nil)

	result, err := service.Create(context.Background(), createRequest)

	assert.NoError(t, err)
	assert.Equal(t, expected, result)
	mockRepo.AssertExpectations(t)
}
{%- endif %}
{%- if "update" in resource.operations %}

func TestUpdate{{ r }}_ShouldReturnUpdated{{ r }}Dto(t *testing.T) {
	mockRepo, service := setup{{ r }}ServiceTest()
	updateRequest := models.Update{{ r }}Request{Id: "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c", Name: "mockUpdate{{ r }}"}
	expected := &models.{{ r }}Dto{Id: "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c", Name: "mockUpdate{{ r }}"}
	mockRepo.On("Update", mock.Anything, mock.AnythingOfType("models.{{ r }}")).Return(expected, nil)

	result, err := service.Update(context.Background(), updateRequest)

	assert.NoError(t, err)
	assert.Equal(t, expected, result)
	mockRepo.AssertExpectations(t)
}
{%- endif %}
{%- if "replace" in resource.operations %}

func TestReplace{{ r }}_ShouldReturnReplaced{{ r }}Dto(t *testing.T) {
	mockRepo, service := setup{{ r }}ServiceTest()
	replaceRequest := models.Replace{{ r }}Request{Id: "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c", Name: "mockReplace{{ r }}"}
	expected := &models.{{ r }}Dto{Id: "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c", Name: "mockReplace{{ r }}"}
	mockRepo.On("Replace", mock.Anything, mock.AnythingOfType("models.{{ r }}")).Return(expected, nil)

	result, err := service.Replace(context.Background(), replaceRequest)

	assert.NoError(t, err)
	assert.Equal(t, expected, result)
	mockRepo.AssertExpectations(t)
}
{%- endif %}
{%- if "delete" in resource.operations %}

func TestDelete{{ r }}_ShouldCallRepositoryDelete(t *testing.T) {
	mockRepo, service := setup{{ r }}ServiceTest()
	id := "0f3a7ff7-a601-4d23-b33c-7f8f18b57a4c"
	mockRepo.On("Delete", mock.Anything, id).Return(nil)

	err := service.Delete(context.Background(), id)

	assert.NoError(t, err)
	mockRepo.AssertCalled(t, "Delete", mock.Anything, id)
}
{%- endif %}
{% endfor %}